eth/consensus/clique/snapshot_manager.py

```diff
@@ -0,0 +1,311 @@
+        if block_hash in self._snapshots:
+            return self._snapshots[block_hash]
+
+        if is_checkpoint(block_number, self._epoch_length):
+            try:
+                # We might have it saved on disk
+                return self.get_snapshot_from_db(block_hash)
+            except SnapshotNotFound:
+                pass
+
+            try:
+                # Otherwise, we can retrieve it on the fly
+                header = self._chain_db.get_block_header_by_hash(block_hash)
+            except HeaderNotFound:
+                raise SnapshotNotFound(f"Can not get snapshot for {block_hash} at {block_number}")
+            else:
+                if header.block_number != block_number:
+                    raise SnapshotNotFound(
+                        f"Can not get snapshot for {block_hash} at {block_number}"
+                    )
+                else:
+                    return self._create_snapshot_from_checkpoint_header(header)
+
+        raise SnapshotNotFound(f"Can not get snapshot for {block_hash} at {block_number}")
+
+    def add_snapshot(self, mutable_snapshot: MutableSnapshot) -> Snapshot:
+        """
+        Retrieve a ``Snapshot`` from the given ``mutable_snapshot`` and add it to the local cache.
+        """
+        snapshot = mutable_snapshot.get_frozen_clone()
+        self._snapshots[snapshot.block_hash] = snapshot
+        return snapshot
+
+    def persist_snapshot(self, snapshot: Snapshot) -> None:
+        """
+        Persist the given snapshot to the database.
+        """
+        key = make_snapshot_lookup_key(snapshot.block_hash)
+        with self._chain_db.db.atomic_batch() as db:
+            db[key] = encode_snapshot(snapshot)
+
+    def get_snapshot_from_db(self, block_hash: Hash32) -> Snapshot:
+        """
+        Retrieve a snapshot from the database. Raise ``SnapshotNotFound`` if it does not exist.
+        """
+        key = make_snapshot_lookup_key(block_hash)
+        try:
+            encoded_key = self._chain_db.db[key]
+        except KeyError as e:
+            raise SnapshotNotFound(
+                f"Can not get on-disk snapshot for {block_hash}"
+            )
+        else:
+            return decode_snapshot(encoded_key)
+
+    def is_valid_vote(self,
+                      signers: Sequence[Address],
+                      subject: Address,
+                      action: VoteAction) -> bool:
+        """
+        Check if a vote is semantically valid, meaning it is either voting against an existing
+        member or voting in favor of an address that is not yet a member.
+        """
+        signer_exists = subject in signers
+        signer_is_kicked = signer_exists and action is VoteAction.KICK
+        signer_is_nominated = not signer_exists and action is VoteAction.NOMINATE
+        return signer_is_kicked or signer_is_nominated
+
+    def cast(self, snapshot: MutableSnapshot, subject: Address, action: VoteAction) -> bool:
```

Comment:
Maybe make this `cast_vote` for more explicit naming.  The bare `cast` had me confused for a moment.